mods/logon-logoff-shutdown-sounds.wh.cpp

```diff
@@ -0,0 +1,582 @@
+// ==WindhawkMod==
+// @id              logon-logoff-shutdown-sounds
+// @name            Logon, Logoff & Shutdown Sounds Restored
+// @description     Restores the logon, logoff and shutdown sounds from earlier versions of Windows
+// @version         1.0.0
+// @author          Toru the Red Fox
+// @github          https://github.com/TorutheRedFox
+// @twitter         https://twitter.com/TorutheRedFox
+// @include         explorer.exe
+// @compilerOptions -lcomdlg32 -lshlwapi -lwinmm -luser32
+// ==/WindhawkMod==
+
+// ==WindhawkModReadme==
+/*
+# Logon Sounds Restored
+Restores the logon, logoff and shutdown sounds from earlier versions of Windows, simple as.
+
+It is recommended to use [these reg files](https://www.howtogeek.com/wp-content/uploads/2016/09/Shutdown-Logoff-Logon-Sound-Hacks.zip) to restore the sound events to the Sound control panel applet.
+*/
+// ==/WindhawkModReadme==
+
+// ==WindhawkModSettings==
+/*
+
+- xpmode: false
+  $name: XP Startup behavior
+  $description: Plays the startup sound on logon. Resuming will still play the logon sound. It is recommended to untick the box to play the startup sound in the control panel to prevent it from playing before you log on.
+
+*/
+// ==/WindhawkModSettings==
+
+#include <gdiplus.h>
+#include <handleapi.h>
+#include <libloaderapi.h>
+#include <processthreadsapi.h>
+#include <windhawk_api.h>
+#include <windhawk_utils.h>
+#include <windows.h>
+#include <shlwapi.h>
+#include <winuser.h>
+
+// settings
+struct {
+    bool    bXpMode;
+} settings;
+
+// the various global handles used 
+
+HMODULE g_hShlwapi;
+HWND g_hwSound;
+HANDLE g_hSoundThread;
+HMODULE g_hExplorer;
+
+// NOTE: sligtly different versions of this exist in...
+//      SHPlaySound() -> shell32
+//      IEPlaySound() -> shdocvw/browseui
+
+STDAPI_(void) ExplorerPlaySound(LPCTSTR pszSound)
+{
+    // note, we have access only to global system sounds here as we use "Apps\.Default"
+    TCHAR szKey[256];
+    wsprintf(szKey, TEXT("AppEvents\\Schemes\\Apps\\.Default\\%s\\.current"), pszSound);
+
+    TCHAR szFileName[MAX_PATH];
+    szFileName[0] = 0;
+    LONG cbSize = sizeof(szFileName);
+
+    // test for an empty string, PlaySound will play the Default Sound if we
+    // give it a sound it cannot find...
+
+    if ((RegQueryValue(HKEY_CURRENT_USER, szKey, szFileName, &cbSize) == ERROR_SUCCESS)
+        && szFileName[0])
+    {
+        // flags are relevant, we try to not stomp currently playing sounds
+        PlaySound(szFileName, NULL, SND_FILENAME | SND_ASYNC | SND_NODEFAULT | SND_NOSTOP);
+    }
+}
+
+typedef LRESULT (* _OnSessionChange_t)(void* _this, WPARAM wParam, LPARAM lParam);
+_OnSessionChange_t _OnSessionChange_orig;
+LRESULT _OnSessionChange_hook(void* _this, WPARAM wParam, LPARAM lParam)
```

Comment:
Calling conventions are missing, which is likely to cause a crash for 32-bit targets. Please add calling conventions or, if the mod doesn't support 32-bit, limit it to 64-bit, which will also make compilation faster. See:
https://github.com/ramensoftware/windhawk/wiki/Development-tips#hooks-and-calling-conventions

Reply:
You only had to add it for hooks, in your case just `_OnSessionChange_hook`, but the additional calling conventions don't harm.